src/coreclr/jit/lower.cpp

```diff
@@ -3644,15 +3644,15 @@ void Lowering::LowerRetStruct(GenTreeUnOp* ret)
     assert(ret->OperIs(GT_RETURN));
     assert(varTypeIsStruct(ret));
 
-    GenTree* retVal = ret->gtGetOp1();
+    GenTree*  retVal           = ret->gtGetOp1();
+    var_types nativeReturnType = comp->info.compRetNativeType;
     // Note: small types are returned as INT.
```

Comment:
Is that comment still correct with your changes?

Reply:
should be still relevant, the following `genActualType` should upscale all small-types to int

Reply:
Right, there are two nodes here, the return and the operand of the return. The return is still wide but the bug was that we were retyping the operand, which is wrong.